### PR TITLE
Don't save exact duplicates in merged index

### PR DESCRIPTION
**Pull request by aawsome**
What does this PR change? What problem does it solve?
-----------------------------------------------------

Don't save exact duplicates (i.e. same blob ID, same pack, same offset and same length) in index. This situation can happen in rare cases where superseded index files are not deleted. Saving exact duplicates doesn't help anywhere but may pose some problems (more memory usage; and possible trouble within #2842) 

Was the change discussed in an issue or in the forum before?
------------------------------------------------------------

see #2839 

Checklist
---------

- [x] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [x] I have enabled [maintainer edits for this PR](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)
- [x] I have added tests for all changes in this PR
- [ ] I have added documentation for the changes (in the manual)
- [ ] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [x] I have run `gofmt` on the code in all commits
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [x] I'm done, this Pull Request is ready for review

### CODE REVIEW

File: internal/repository/indexmap.go [hunk 1]
```diff
@@ -34,7 +34,7 @@ const (
 )
 
 // add inserts an indexEntry for the given arguments into the map,
-// using id as the key.
+// does not insert duplicates (all values identical to an existing one).
```
Comment (greatroar) [inline internal/repository/indexmap.go:37]: This is a new sentence and it contradicts the comment on type indexMap.

```
 // add inserts an indexEntry for the given arguments into the map.
// It allows duplicate ids, but not fully duplicate entries.
```

The comment on indexMap needs an update too.

**Comment by greatroar** [comment]
I haven't investigated the problem, but the code looks good. Insertion will be more expensive, but is still O(1) expected. Merge reduces to add so that is covered as well (maybe add a test for that?).

File: internal/repository/indexmap_test.go [hunk 2]
```diff
@@ -44,26 +44,38 @@ func TestIndexMapForeach(t *testing.T) {
...
+	testForEach := func() {
+		seen := make(map[int]struct{})
+		m.foreach(func(e *indexEntry) bool {
+			i := int(e.id[0])
+			rtest.Assert(t, i < N, "unknown id %v in indexMap", e.id)
+			rtest.Equals(t, i, e.packIndex)
+			rtest.Equals(t, i, int(e.length))
+			rtest.Equals(t, i, int(e.offset))
+
+			seen[i] = struct{}{}
```
Comment (MichaelEischer) [inline internal/repository/indexmap_test.go:56]: This doesn't count duplicate ids. Please make sure that your tests actually _fail_ without the fixed code. So either write the test first or undo the change afterwards and check that the test fails.

Comment (aawsome) [reply internal/repository/indexmap_test.go:56]: My bad :-1: 
I'll change this.

**Comment by aawsome** [comment]
There is another issue I just came around when thinking about this change once again: The packIDs are not deduplicated so far.. 
I need to think about how to best do this and change this to WIP in meanwhile..

**Comment by aawsome** [comment]
I changed this such that now the check is performed in `Index.merge()`. IndexMap` stays unchanged now.

File: internal/repository/index.go [hunk 3]
```diff
@@ -476,18 +476,30 @@ func (idx *Index) merge(idx2 *Index) error {
...
+
 	// copy all index entries of idx2 to idx
 	for typ := range idx2.byType {
 		m2 := &idx2.byType[typ]
 		m := &idx.byType[typ]
-		m2.foreach(func(entry *indexEntry) bool {
-			// packIndex is changed as idx2.pack is appended to idx.pack, see below
-			m.add(entry.id, entry.packIndex+packlen, entry.offset, entry.length)
+		m2.foreach(func(e2 *indexEntry) bool {
+			found := false
```
Comment (MichaelEischer) [inline internal/repository/index.go:487]: Please add a comment the the `merge` method which mentions that duplicate entries are removed.

The next few lines also needs a comment. And I'd prefer these to be move into a method like `hasPackedBlob` or similar. At first glance it's not obvious at all what the foreach inside another foreach does.

File: internal/repository/master_index_test.go [hunk 4]
```diff
@@ -167,6 +168,14 @@ func TestMasterMergeFinalIndexes(t *testing.T) {
 	rtest.Equals(t, []*repository.Index{idx1, idx2}, finalIndexes)
 
 	mIdx.MergeFinalIndexes()
+	allIndexes := mIdx.All()
+	rtest.Equals(t, 1, len(allIndexes))
+
+	count := 0
```
Comment (MichaelEischer) [inline internal/repository/master_index_test.go:174]: Please rename the variable to `blobCount`.

**Review by MichaelEischer — COMMENTED** [review]
@aawsome Did you check the masterIndex benchmarks regarding the performance impact of the additional `foreach` loop?

File: internal/repository/index.go (hunk 3 above)
Comment (aawsome) [reply internal/repository/index.go]: is changed.

File: internal/repository/master_index_test.go (hunk 4 above)
Comment (aawsome) [reply internal/repository/master_index_test.go]: is changed

**Comment by aawsome** [comment]
> @aawsome Did you check the masterIndex benchmarks regarding the performance impact of the additional `foreach` loop?

Note that the additional `foreachID` loop is `O(1)` itself (at least if the undelying hash is not degenerated). Hence the total complexity of the merge is still roughly (neglecting hash growing of `idx`) `O(n)` with `n` being the number of elements of the index to be merged, i.e. `idx2`. 

Moreover this change only affects the merge itself, and in real-world-scenarios this occurs during 

- loading the index from index files into memory (should be dominated by loading and extracting the files)
- saving a new finished index during backup (this should happen quite rarely as backup itself should be mainly dominated by 

So I do not expect any impact on real-life usage.

About benchmarking: Seems to me that none of the current benchmarks actually includes merging of indexes in the measurements.
I added the benchmark `MasterIndexAlloc` which creates Indexes and merges them under measurement.
Here are the results comparing 5e63294 with b112533:
```
name                old time/op    new time/op    delta
MasterIndexAlloc-4     334ms ± 1%     460ms ± 1%  +37.68%  (p=0.000 n=9+8)

name                old alloc/op   new alloc/op   delta
MasterIndexAlloc-4     258MB ± 0%     258MB ± 0%     ~     (p=0.101 n=10+8)

name                old allocs/op  new allocs/op  delta
MasterIndexAlloc-4      317k ± 0%      317k ± 0%     ~     (p=0.267 n=10+6)
```

**Review by MichaelEischer — APPROVED** [review]
LGTM.

> Note that the additional `foreachID` loop is `O(1)` itself (at least if the undelying hash is not degenerated). Hence the total complexity of the merge is still roughly (neglecting hash growing of `idx`) `O(n)` with `n` being the number of elements of the index to be merged, i.e. `idx2`.

`foreachID` could end up a rather large value if a single blob exists lots of times in different packs. But in that case the repository is probably really broken. So `amortized O(1)` would be slightly more precise (not that it matters).

The overhead of 40% seems fine for me, I was worried a bit that we could end up with an overhead of maybe 2x-3x.

**Pull request merged**